### PR TITLE
[jit] Fix inspect.get_annotations usage in python >= 3.10

### LINKED ISSUE #104484: [JIT] inherited scripting can fail with parent class annotations in python >= 3.10
### 🐛 Describe the bug

```python
import sys
import abc
from typing import List
import torch
import inspect

class BaseModule(torch.nn.Module, metaclass=abc.ABCMeta):
    state: List[int]

    @abc.abstractmethod
    def forward(self, x):
        """
        docstring
        """

def do_something_with_list(x: List[int]):
    if x:
        return x[-1]
    return 5

class Submodule(BaseModule):
    def __init__(self, self_x_value):
        super().__init__()
        self.x = self_x_value
        self.state = []

    def forward(self, x):
        return self.x + x + do_something_with_list(self.state)

class LowestModule(Submodule):
    def __init__(self):
        super().__init__(123)

mod = LowestModule()
mod2 = LowestModule()
torch.jit.script(mod)
torch.jit.script(mod2)
```

Error:

```
Traceback (most recent call last):
  File "/home/dberard/repro_gh.py", line 37, in <module>
    torch.jit.script(mod2)
  File "/data/users/dberard/pytorch310/pytorch/torch/jit/_script.py", line 1284, in script
    return torch.jit._recursive.create_script_module(
  File "/data/users/dberard/pytorch310/pytorch/torch/jit/_recursive.py", line 480, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/data/users/dberard/pytorch310/pytorch/torch/jit/_recursive.py", line 546, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/data/users/dberard/pytorch310/pytorch/torch/jit/_recursive.py", line 397, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
RuntimeError:
Module 'LowestModule' has no attribute 'state' (This attribute exists on the Python module, but we failed to convert Python type: 'list' to a TorchScript type. List trace inputs must have elements. Its type was inferred; try adding a type annotation for the attribute.):
  File "/home/dberard/repro_gh.py", line 28
    def forward(self, x):
        return self.x + x + do_something_with_list(self.state)
                                                   ~~~~~~~~~~ <--- HER
```

Note, this appears to be due to some weird behavior in inspect.get_annotations:

```python
import sys
import abc
from typing import List
import torch
import inspect

class BaseModule(torch.nn.Module, metaclass=abc.ABCMeta):
    state: List[int]

    @abc.abstractmethod
    def forward(self, x):
        """
        docstring
        """

def do_something_with_list(x: List[int]):
    if x:
        return x[-1]
    return 5

class Submodule(BaseModule):
    def __init__(self, self_x_value):
        super().__init__()
        self.x = self_x_value
        self.state = []

    def forward(self, x):
        return self.x + x + do_something_with_list(self.state)

class LowestModule(Submodule):
    def __init__(self):
        super().__init__(123)

mod = LowestModule()
print(inspect.get_annotations(mod))  # prints {'state': List[int]}
inspect.getmembers(type(mod))
print(inspect.get_annotations(mod))  # prints {}
```

### Versions

pytorch main branch, python 3.10

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104485

Fixes #104484

For >= 3.10, we use `inspect.get_annotations` instead of `getattr(.., "__annotations__")`. [Docs](https://docs.python.org/3/library/inspect.html#inspect.get_annotations) say that get_annotations() "Ignores inherited annotations on classes. If a class doesn’t have its own annotations dict, returns an empty dict.". In practice though, this doesn't seem always true; until you call inspect.getmembers it seems like you still get inherited annotations. In particular, this means that if you script a certain type twice, the first time it may pass scripting but on the second try it may not pass scripting.

This PR adds a more comprehensive handling of get_annotations by recursively reading the annotations of the base types. (TorchScript doesn't officially support this; but since it worked in <3.10, it's now breaking internal stuff as python gets upgraded to 3.10)

Verified in #104486 that the test does actually fail before the changes in this PR were added.

Differential Revision: [D47163891](https://our.internmc.facebook.com/intern/diff/D47163891)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47163891).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47163891).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47163891).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47163891).

**Review by eellison — APPROVED** [review]
niceeeee

File: torch/jit/_recursive.py [hunk 1]
```diff
@@ -149,12 +149,21 @@ def get_annotations(obj):
...
+        if annotations:
             return annotations
-        cls = obj if isinstance(obj, type) else type(obj)
-        if len(cls.__bases__) == 0:
+
+        def get_cls_annotations(cls):
+            cls_annotations = inspect.get_annotations(cls)
+            if cls_annotations:
+                return cls_annotations
+            for base in cls.__bases__:
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/jit/_recursive.py:159]: I did some experiments with `mod.__annotations__` and I think is the implementation that best matches `mod.__annotations__`.

**Pull request closed without merging**